torch/csrc/jit/passes/symbolic_shape_analysis.cpp

```diff
@@ -0,0 +1,202 @@
+
+c10::optional<size_t> normIndex(int64_t index, size_t len) {
+  if (index < 0) {
+    index = index + len;
+  }
+  if (index >= 0 && index < static_cast<int64_t>(len)) {
+    return index;
+  } else {
+    return c10::nullopt;
+  }
+}
+
+void replaceWithIValue(Value* v, IValue val) {
+  WithInsertPoint guard(*v->node()->owningBlock()->nodes().begin());
+  v->replaceAllUsesWith(v->owningGraph()->insertConstant(val));
+}
+
+// Symbolic Shape Analysis works through iteratively partially evaluating
+// a TorchScript shape compute graph by inputting properties from input
+// Tensors. We can substitute in properties like `len(x)` and `x[1]`
+// if they are statically on the input Tensors. We can also use
+// assertions like `assert len(x) == 4` in order to refine the input
+// length and unroll loops over its elements. We iterativley optimize and
+// substitute in properties until we are unable to make any further
+// optimizations. Finally, we try to extract Tensor properties from the output.
+// For instance `return [1, 2, inp[2] + 1, inp[3]]` we know that the ouptut
+// will be length 4 with first two dimensions equal to 1 and 2.
+// It is not implemented yet but in the future we will also be able to
+// infer that the 4th dimension will have the same symbolic shape as inp[3]
+
+struct SymbolicShapeAnalyzer {
+  SymbolicShapeAnalyzer(std::shared_ptr<Graph> graph, Node* n)
+      : graph_(graph->copy()), node_(n) {
+    for (size_t i = 0; i < node_->inputs().size(); i++) {
+      auto type = node_->input(i)->type();
+      if (auto tt = type->castRaw<TensorType>()) {
+        c10::SymbolicShape symbolic_shapes = tt->symbolic_sizes();
+        if (symbolic_shapes.isComplete()) {
+          replaceWithIValue(
+              graph_->inputs().at(i), *tt->sizes().concrete_sizes());
+          continue;
+        }
+        tensor_indices.push_back(i);
+      } else if (
+          type->cast<ListType>() &&
+          type->cast<ListType>()->getElementType()->cast<TensorType>()) {
+        TORCH_INTERNAL_ASSERT(false); // not handled yet
+      } else {
+        if (auto ival = toIValue(node_->input(i))) {
+          replaceWithIValue(graph_->inputs().at(i), *ival);
+        }
+      }
+    }
+  }
+
+  c10::SymbolicShape run() {
+    // TODO: only run while the last iteration has made a change
+    for (size_t i = 0; i < 6; i++) {
+      substituteInputTensorProperties();
+      LowerSimpleTuples(graph_);
+      RemoveListMutation(graph_);
+      UnrollConstantLoops(graph_);
+      ConstantPropagation(graph_);
+      PeepholeOptimize(graph_);
+      ConstantPropagation(graph_);
+    }
+    ConstantPooling(graph_);
+    EliminateDeadCode(graph_);
+    return extractOutputShape();
+  }
+
+ private:
+  void substituteInputTensorProperties() {
+    for (auto index : tensor_indices) {
+      substituteTensorProperties(index);
+    }
+  }
+
+  void substituteTensorProperties(int64_t tensor_index) {
```

Comment:
Nit: I find it a bit confusing that we use different terms here and there without defining what they mean: tensor_indices, dims, shape, tensor_value. What's the difference between all of them? Could we add some comments?